File: src/main/java/com/jigumulmi/admin/place/AdminPlaceController.java
package com.jigumulmi.admin.place;

import com.jigumulmi.admin.place.dto.WeeklyBusinessHourDto;
import com.jigumulmi.admin.place.dto.request.AdminCreatePlaceRequestDto;
import com.jigumulmi.admin.place.dto.request.AdminCreateTemporaryBusinessHourRequestDto;
import com.jigumulmi.admin.place.dto.request.AdminGetPlaceListRequestDto;
import com.jigumulmi.admin.place.dto.response.AdminCreatePlaceResponseDto;
import com.jigumulmi.admin.place.dto.response.AdminPlaceBasicResponseDto;
import com.jigumulmi.admin.place.dto.response.AdminPlaceBusinessHourResponseDto;
import com.jigumulmi.admin.place.dto.response.AdminPlaceListResponseDto.PlaceDto;
import com.jigumulmi.admin.place.dto.response.AdminS3DeletePresignedUrlResponseDto;
import com.jigumulmi.admin.place.dto.response.AdminS3PutPresignedUrlResponseDto;
import com.jigumulmi.common.PagedResponseDto;
import com.jigumulmi.config.security.RequiredAuthUser;
import com.jigumulmi.member.domain.Member;
import com.jigumulmi.place.dto.ImageDto;
import com.jigumulmi.place.dto.MenuDto;
import com.jigumulmi.place.dto.request.MenuImageS3DeletePresignedUrlRequestDto;
import com.jigumulmi.place.dto.request.MenuImageS3PutPresignedUrlRequestDto;
import com.jigumulmi.place.dto.response.DistrictResponseDto;
import com.jigumulmi.place.vo.Region;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.media.ArraySchema;
import io.swagger.v3.oas.annotations.media.Content;
import io.swagger.v3.oas.annotations.media.Schema;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.responses.ApiResponses;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import java.time.LocalDate;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.springdoc.core.annotations.ParameterObject;
import org.springframework.data.domain.Pageable;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.ModelAttribute;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

@Tag(name = "장소 관리")
@RestController
@RequiredArgsConstructor
@RequestMapping("/admin/place")
public class AdminPlaceController {

    private final AdminPlaceService adminPlaceService;


    @Operation(summary = "장소 리스트 조회")
    @GetMapping("")
    public ResponseEntity<PagedResponseDto<PlaceDto>> getPlaceList(
        @ParameterObject Pageable pageable,
        @ModelAttribute AdminGetPlaceListRequestDto requestDto) {
        PagedResponseDto<PlaceDto> placeList = adminPlaceService.getPlaceList(pageable, requestDto);
        return ResponseEntity.ok().body(placeList);
    }

    @Operation(summary = "장소 기본 정보 조회")
    @GetMapping("/{placeId}/basic")
    public ResponseEntity<AdminPlaceBasicResponseDto> getPlaceBasic(@PathVariable Long placeId) {
        AdminPlaceBasicResponseDto responseDto = adminPlaceService.getPlaceBasic(placeId);
        return ResponseEntity.ok().body(responseDto);
    }

    @Operation(summary = "장소 기본 정보 수정", description = "덮어쓰는 로직이므로 수정되지 않은 기존 데이터도 필요")
    @ApiResponse(responseCode = "201")
    @PutMapping("/{placeId}/basic")
    public ResponseEntity<?> updatePlaceBasic(@PathVariable Long placeId,
        @RequestBody AdminCreatePlaceRequestDto requestDto) {
        adminPlaceService.updatePlaceBasic(placeId, requestDto);
        return ResponseEntity.status(HttpStatus.CREATED).build();
    }

    @Operation(summary = "장소 이미지 조회")
    @GetMapping("/{placeId}/image")
    public ResponseEntity<List<ImageDto>> getPlaceImage(@PathVariable Long placeId) {
        List<ImageDto> responseDto = adminPlaceService.getPlaceImage(placeId);
        return ResponseEntity.ok().body(responseDto);
    }

    @Operation(summary = "장소 이미지 수정", description = "덮어쓰는 로직이므로 수정되지 않은 기존 데이터도 필요")
    @ApiResponse(responseCode = "201")
    @PutMapping("/{placeId}/image")
    public ResponseEntity<?> updatePlaceImage(@PathVariable Long placeId,
        @RequestBody List<ImageDto> requestDto) {
        adminPlaceService.updatePlaceImage(placeId, requestDto);
        return ResponseEntity.status(HttpStatus.CREATED).build();
    }

    @Operation(summary = "장소 메뉴 조회")
    @GetMapping("/{placeId}/menu")
    public ResponseEntity<List<MenuDto>> getMenu(@PathVariable Long placeId) {
        List<MenuDto> responseDto = adminPlaceService.getMenu(placeId);
        return ResponseEntity.ok().body(responseDto);
    }

    @Operation(summary = "장소 메뉴 수정", description = "덮어쓰는 로직이므로 수정되지 않은 기존 데이터도 필요")
    @ApiResponse(responseCode = "201")
    @PutMapping("/{placeId}/menu")
    public ResponseEntity<?> updateMenu(@PathVariable Long placeId,
        @RequestBody List<MenuDto> requestDto) {
        adminPlaceService.updateMenu(placeId, requestDto);
        return ResponseEntity.status(HttpStatus.CREATED).build();
    }

    @Operation(summary = "장소 고정 영업시간 수정", description = "덮어쓰는 로직이므로 수정되지 않은 기존 데이터도 필요")
    @ApiResponse(responseCode = "201")
    @PutMapping("/{placeId}/business-hour/fixed")
    public ResponseEntity<?> updateFixedBusinessHour(@PathVariable Long placeId,
        @Valid @RequestBody WeeklyBusinessHourDto requestDto) {
        adminPlaceService.updateFixedBusinessHour(placeId, requestDto);
        return ResponseEntity.status(HttpStatus.CREATED).build();
    }

    @Operation(summary = "장소 변동 영업시간 생성")
    @ApiResponse(responseCode = "201")
    @PostMapping("/{placeId}/business-hour/temporary")
    public ResponseEntity<?> createTemporaryBusinessHour(@PathVariable Long placeId,
        @Valid @RequestBody AdminCreateTemporaryBusinessHourRequestDto requestDto) {
        adminPlaceService.createTemporaryBusinessHour(placeId, requestDto);
        return ResponseEntity.status(HttpStatus.CREATED).build();
    }

    @Operation(summary = "장소 변동 영업시간 수정", description = "덮어쓰는 로직이므로 수정되지 않은 기존 데이터도 필요, 날짜만 변경도 가능")
    @ApiResponse(responseCode = "201")
    @PutMapping("/{placeId}/business-hour/temporary/{temporaryBusinessHourId}")
    public ResponseEntity<?> updateTemporaryBusinessHour(@PathVariable Long placeId,
        @PathVariable Long temporaryBusinessHourId,
        @Valid @RequestBody AdminCreateTemporaryBusinessHourRequestDto requestDto) {
        adminPlaceService.updateTemporaryBusinessHour(temporaryBusinessHourId, requestDto);
        return ResponseEntity.status(HttpStatus.CREATED).build();
    }

    @Operation(summary = "장소 변동 영업시간 삭제")
    @ApiResponse(responseCode = "204")
    @DeleteMapping("/{placeId}/business-hour/temporary/{temporaryBusinessHourId}")
    public ResponseEntity<?> deleteTemporaryBusinessHour(@PathVariable Long placeId,
        @PathVariable Long temporaryBusinessHourId) {
        adminPlaceService.deleteTemporaryBusinessHour(temporaryBusinessHourId);
        return ResponseEntity.noContent().build();
    }

    @Operation(summary = "장소 영업시간 조회")
    @GetMapping("/{placeId}/business-hour")
    public ResponseEntity<AdminPlaceBusinessHourResponseDto> getPlaceBusinessHour(
        @PathVariable Long placeId, @RequestParam(required = false) Integer month) {
        month = (month == null) ? LocalDate.now().getMonthValue() : month;
        AdminPlaceBusinessHourResponseDto responseDto = adminPlaceService.getPlaceBusinessHour(
            placeId, month);
        return ResponseEntity.ok().body(responseDto);
    }

    @Operation(summary = "장소 생성")
    @ApiResponse(responseCode = "201")
    @PostMapping("")
    public ResponseEntity<AdminCreatePlaceResponseDto> createPlace(
        @RequestBody AdminCreatePlaceRequestDto requestDto,
        @RequiredAuthUser Member member
    ) {
        AdminCreatePlaceResponseDto responseDto = adminPlaceService.createPlace(requestDto, member);
        return ResponseEntity.status(HttpStatus.CREATED).body(responseDto);
    }

    @Operation(summary = "장소 삭제")
    @ApiResponse(responseCode = "204")
    @DeleteMapping("/{placeId}")
    public ResponseEntity<?> deletePlace(@PathVariable Long placeId) {
        adminPlaceService.deletePlace(placeId);
        return ResponseEntity.noContent().build();
    }

    @Operation(summary = "광역시도 조회")
    @ApiResponses(
        value = {@ApiResponse(responseCode = "200", content = {
            @Content(array = @ArraySchema(schema = @Schema(implementation = Region.class)))})}
    )
    @GetMapping("/region")
    public ResponseEntity<?> getRegionList() {
        List<Region> regionList = adminPlaceService.getRegionList();
        return ResponseEntity.ok().body(regionList);
    }

    @Operation(summary = "시군구 조회")
    @ApiResponses(
        value = {@ApiResponse(responseCode = "200", content = {
            @Content(array = @ArraySchema(schema = @Schema(implementation = DistrictResponseDto.class)))})}
    )
    @GetMapping("/district")
    public ResponseEntity<?> getDistrictList(@RequestParam Region region) {
        List<DistrictResponseDto> districtList = adminPlaceService.getDistrictList(region);
        return ResponseEntity.ok().body(districtList);
    }

    @Operation(summary = "메뉴 이미지 S3 Put Presigned Url 요청")
    @ApiResponse(responseCode = "201", content = {
        @Content(schema = @Schema(implementation = AdminS3PutPresignedUrlResponseDto.class))})
    @PostMapping("/menu/s3-put-presigned-url")
    public ResponseEntity<?> createMenuImageS3PutPresignedUrl(
        @Valid @RequestBody MenuImageS3PutPresignedUrlRequestDto requestDto,
        @RequiredAuthUser Member Member) {
        AdminS3PutPresignedUrlResponseDto responseDto = adminPlaceService.createMenuImageS3PutPresignedUrl(
            requestDto);
        return ResponseEntity.status(HttpStatus.CREATED).body(responseDto);
    }

    @Operation(summary = "메뉴 이미지 S3 Delete Presigned Url 요청")
    @ApiResponse(responseCode = "201", content = {
        @Content(schema = @Schema(implementation = AdminS3DeletePresignedUrlResponseDto.class))})
    @PostMapping("/menu/s3-delete-presigned-url")
    public ResponseEntity<?> createMenuImageS3DeletePresignedUrl(
        @Valid @RequestBody MenuImageS3DeletePresignedUrlRequestDto requestDto,
        @RequiredAuthUser Member Member) {
        AdminS3DeletePresignedUrlResponseDto responseDto = adminPlaceService.createMenuImageS3DeletePresignedUrl(
            requestDto);
        return ResponseEntity.status(HttpStatus.CREATED).body(responseDto);
    }
}

File: admin/src/main/kotlin/com/jigumulmi/place/AdminPlaceController.kt
package com.jigumulmi.place

import com.jigumulmi.common.PagedResponseDto
import com.jigumulmi.config.security.RequiredAuthUser
import com.jigumulmi.member.domain.Member
import com.jigumulmi.place.dto.ImageDto
import com.jigumulmi.place.dto.MenuDto
import com.jigumulmi.place.dto.WeeklyBusinessHourDto
import com.jigumulmi.place.dto.request.*
import com.jigumulmi.place.dto.response.*
import com.jigumulmi.place.dto.response.AdminPlaceListResponseDto.PlaceDto
import com.jigumulmi.place.vo.Region
import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.responses.ApiResponse
import io.swagger.v3.oas.annotations.responses.ApiResponses
import io.swagger.v3.oas.annotations.tags.Tag
import jakarta.validation.Valid
import org.springdoc.core.annotations.ParameterObject
import org.springframework.data.domain.Pageable
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.time.LocalDate

@Tag(name = "장소 관리")
@RestController
@RequestMapping("/admin/place")
class AdminPlaceController(
    private val adminPlaceService: AdminPlaceService
) {

    @Operation(summary = "장소 리스트 조회")
    @GetMapping("")
    fun getPlaceList(
        @ParameterObject pageable: Pageable,
        @ModelAttribute requestDto: AdminGetPlaceListRequestDto
    ): ResponseEntity<PagedResponseDto<PlaceDto>> {
        val placeList = adminPlaceService.getPlaceList(pageable, requestDto)
        return ResponseEntity.ok().body(placeList)
    }

    @Operation(summary = "장소 기본 정보 조회")
    @GetMapping("/{placeId}/basic")
    fun getPlaceBasic(@PathVariable placeId: Long): ResponseEntity<AdminPlaceBasicResponseDto> {
        val responseDto = adminPlaceService.getPlaceBasic(placeId)
        return ResponseEntity.ok().body(responseDto)
    }

    @Operation(summary = "장소 기본 정보 수정", description = "덮어쓰는 로직이므로 수정되지 않은 기존 데이터도 필요")
    @ApiResponses(
        ApiResponse(responseCode = "201"),
        ApiResponse(responseCode = "400", description = "승인된 장소 수정 시 데이터 누락")
    )
    @PutMapping("/{placeId}/basic")
    fun updatePlaceBasic(
        @PathVariable placeId: Long,
        @RequestBody requestDto: AdminCreatePlaceRequestDto
    ): ResponseEntity<*> {
        requestDto.validate()
        adminPlaceService.updatePlaceBasic(placeId, requestDto)
        return ResponseEntity.status(HttpStatus.CREATED).build<Any>()
    }

    @Operation(summary = "장소 이미지 조회")
    @GetMapping("/{placeId}/image")
    fun getPlaceImage(@PathVariable placeId: Long): ResponseEntity<List<ImageDto?>?> {
        val responseDto = adminPlaceService.getPlaceImage(placeId)
        return ResponseEntity.ok().body(responseDto)
    }

    @Operation(summary = "장소 이미지 수정", description = "덮어쓰는 로직이므로 수정되지 않은 기존 데이터도 필요")
    @ApiResponses(
        ApiResponse(responseCode = "201"),
        ApiResponse(responseCode = "400", description = "승인된 장소 수정 시 데이터 누락")
    )
    @PutMapping("/{placeId}/image")
    fun updatePlaceImage(
        @PathVariable placeId: Long,
        @RequestBody requestDto: List<ImageDto>
    ): ResponseEntity<*> {
        adminPlaceService.updatePlaceImage(placeId, requestDto)
        return ResponseEntity.status(HttpStatus.CREATED).build<Any>()
    }

    @Operation(summary = "장소 메뉴 조회")
    @GetMapping("/{placeId}/menu")
    fun getMenu(@PathVariable placeId: Long): ResponseEntity<List<MenuDto>> {
        val responseDto = adminPlaceService.getMenu(placeId)
        return ResponseEntity.ok().body(responseDto)
    }

    @Operation(summary = "장소 메뉴 수정", description = "덮어쓰는 로직이므로 수정되지 않은 기존 데이터도 필요")
    @ApiResponses(
        ApiResponse(responseCode = "201"),
        ApiResponse(responseCode = "400", description = "승인된 장소 수정 시 데이터 누락")
    )
    @PutMapping("/{placeId}/menu")
    fun updateMenu(
        @PathVariable placeId: Long,
        @RequestBody requestDto: List<MenuDto>
    ): ResponseEntity<*> {
        adminPlaceService.updateMenu(placeId, requestDto)
        return ResponseEntity.status(HttpStatus.CREATED).build<Any>()
    }

    @Operation(summary = "장소 고정 영업시간 수정", description = "덮어쓰는 로직이므로 수정되지 않은 기존 데이터도 필요")
    @ApiResponse(responseCode = "201")
    @PutMapping("/{placeId}/business-hour/fixed")
    fun updateFixedBusinessHour(
        @PathVariable placeId: Long,
        @RequestBody requestDto: @Valid WeeklyBusinessHourDto
    ): ResponseEntity<*> {
        adminPlaceService.updateFixedBusinessHour(placeId, requestDto)
        return ResponseEntity.status(HttpStatus.CREATED).build<Any>()
    }

    @Operation(summary = "장소 변동 영업시간 생성")
    @ApiResponse(responseCode = "201")
    @PostMapping("/{placeId}/business-hour/temporary")
    fun createTemporaryBusinessHour(
        @PathVariable placeId: Long,
        @RequestBody requestDto: @Valid AdminCreateTemporaryBusinessHourRequestDto
    ): ResponseEntity<*> {
        adminPlaceService.createTemporaryBusinessHour(placeId, requestDto)
        return ResponseEntity.status(HttpStatus.CREATED).build<Any>()
    }

    @Operation(summary = "장소 변동 영업시간 수정", description = "덮어쓰는 로직이므로 수정되지 않은 기존 데이터도 필요, 날짜만 변경도 가능")
    @ApiResponse(responseCode = "201")
    @PutMapping("/{placeId}/business-hour/temporary/{temporaryBusinessHourId}")
    fun updateTemporaryBusinessHour(
        @PathVariable placeId: Long,
        @PathVariable temporaryBusinessHourId: Long,
        @RequestBody requestDto: @Valid AdminCreateTemporaryBusinessHourRequestDto
    ): ResponseEntity<*> {
        adminPlaceService.updateTemporaryBusinessHour(temporaryBusinessHourId, requestDto)
        return ResponseEntity.status(HttpStatus.CREATED).build<Any>()
    }

    @Operation(summary = "장소 변동 영업시간 삭제")
    @ApiResponse(responseCode = "204")
    @DeleteMapping("/{placeId}/business-hour/temporary/{temporaryBusinessHourId}")
    fun deleteTemporaryBusinessHour(
        @PathVariable placeId: Long,
        @PathVariable temporaryBusinessHourId: Long
    ): ResponseEntity<*> {
        adminPlaceService.deleteTemporaryBusinessHour(temporaryBusinessHourId)
        return ResponseEntity.noContent().build<Any>()
    }

    @Operation(summary = "장소 영업시간 조회")
    @GetMapping("/{placeId}/business-hour")
    fun getPlaceBusinessHour(
        @PathVariable placeId: Long, @RequestParam(required = false) month: Int? = null
    ): ResponseEntity<AdminPlaceBusinessHourResponseDto> {
        val adjustedMonth = if ((month == null)) LocalDate.now().monthValue else month
        val responseDto = adminPlaceService.getPlaceBusinessHour(placeId, adjustedMonth)
        return ResponseEntity.ok().body(responseDto)
    }

    @Operation(summary = "장소 생성")
    @ApiResponse(responseCode = "201")
    @PostMapping("")
    fun createPlace(
        @RequestBody requestDto: AdminCreatePlaceRequestDto,
        @RequiredAuthUser member: Member
    ): ResponseEntity<AdminCreatePlaceResponseDto> {
        requestDto.validate()
        val responseDto = adminPlaceService.createPlace(requestDto, member)
        return ResponseEntity.status(HttpStatus.CREATED).body(responseDto)
    }

    @Operation(summary = "장소 승인/미승인")
    @ApiResponses(
        ApiResponse(responseCode = "204", description = "장소 승인 여부 업데이트 성공"),
        ApiResponse(responseCode = "400", description = "장소 승인 검증 실패")
    )
    @PostMapping("/{placeId}/approval")
    fun togglePlaceApprove(
        @PathVariable placeId: Long,
        @RequestBody requestDto: @Valid TogglePlaceApproveRequestDto
    ): ResponseEntity<*> {
        adminPlaceService.togglePlaceApprove(placeId, requestDto.approve)
        return ResponseEntity.noContent().build<Any>()
    }

    @Operation(summary = "장소 삭제")
    @ApiResponse(responseCode = "204")
    @DeleteMapping("/{placeId}")
    fun deletePlace(@PathVariable placeId: Long): ResponseEntity<*> {
        adminPlaceService.deletePlace(placeId)
        return ResponseEntity.noContent().build<Any>()
    }

    @GetMapping("/region")
    @ApiResponses(value = [ApiResponse(responseCode = "200")])
    @Operation(summary = "광역시도 조회")
    fun getRegionList(): ResponseEntity<List<Region>> {
        val regionList = adminPlaceService.getRegionList()
        return ResponseEntity.ok().body(regionList)
    }

    @Operation(summary = "시군구 조회")
    @ApiResponses(value = [ApiResponse(responseCode = "200")])
    @GetMapping("/district")
    fun getDistrictList(@RequestParam region: Region): ResponseEntity<List<DistrictResponseDto>> {
        val districtList = adminPlaceService.getDistrictList(region)
        return ResponseEntity.ok().body(districtList)
    }

    @Operation(summary = "메뉴 이미지 S3 Put Presigned Url 요청")
    @ApiResponse(responseCode = "201")
    @PostMapping("/menu/s3-put-presigned-url")
    fun createMenuImageS3PutPresignedUrl(
        @RequestBody requestDto: @Valid MenuImageS3PutPresignedUrlRequestDto,
        @RequiredAuthUser Member: Member
    ): ResponseEntity<AdminS3PutPresignedUrlResponseDto> {
        val responseDto = adminPlaceService.createMenuImageS3PutPresignedUrl(
            requestDto
        )
        return ResponseEntity.status(HttpStatus.CREATED).body(responseDto)
    }

    @Operation(summary = "메뉴 이미지 S3 Delete Presigned Url 요청")
    @ApiResponse(responseCode = "201")
    @PostMapping("/menu/s3-delete-presigned-url")
    fun createMenuImageS3DeletePresignedUrl(
        @RequestBody requestDto: @Valid MenuImageS3DeletePresignedUrlRequestDto,
        @RequiredAuthUser Member: Member
    ): ResponseEntity<AdminS3DeletePresignedUrlResponseDto> {
        val responseDto = adminPlaceService.createMenuImageS3DeletePresignedUrl(
            requestDto
        )
        return ResponseEntity.status(HttpStatus.CREATED).body(responseDto)
    }
}
